### PR TITLE
Crypto natives pt 1 cost params: ed25519 and hash

### PR DESCRIPTION
**Pull request by oxade**
## Description 

Adds the cost params logic for natives used in `ed25519.move` and `hash.move`

## Test Plan 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-protocol-config/src/lib.rs [hunk 1]
```diff
@@ -804,6 +839,16 @@ impl ProtocolConfig {
...
                 event_emit_cost_per_byte: Some(1_000),
+                // Crypto
+                // ed25519
+                ed25519_ed25519_verify_cost_base: Some(52),
+                ed25519_ed25519_verify_msg_cost_per_byte: Some(0),
+                // hash::blake2b256
+                hash_blake2b256_cost_base: Some(52),
+                hash_blake2b256_data_cost_per_byte: Some(0),
+                // hash::keccak256
+                hash_keccak256_cost_base: Some(52),
```
Comment (oxade) [inline crates/sui-protocol-config/src/lib.rs:850]: lease remember that the cost numbers are just placeholders right now. We’ll finetune.

File: crates/sui-framework/src/natives/crypto/hash.rs [hunk 2]
```diff
@@ -1,50 +1,116 @@
...
         )],
     ))
 }
 
+#[derive(Clone)]
+pub struct HashKeccak256CostParams {
+    /// Base cost for invoking the `blake2b256` function
+    pub hash_keccak256_cost_base: InternalGas,
+    /// Cost per byte of `data`
+    pub hash_keccak256_data_cost_per_byte: InternalGas,
```
Comment (benr-ml) [inline crates/sui-framework/src/natives/crypto/hash.rs:49]: Hashing 2 bytes or 3 costs the same - Should we change this to cost_per_block? or do we include in it also the move to Rust deserialization and thus we better stay with message length?

Comment (oxade) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: I'm open to either as it's maintainable and best represent how the cost scales with input length.

Comment (benr-ml) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: Yes, Keccak256 has a block size of 32 bytes, so maybe have here:
- hash_keccak256_data_cost_per_byte - to represent the cost of deserialization between Move and Rust
- hash_keccak256_data_cost_per_block - to represent the actual hash operation (using msg.len()/32*hash_keccak256_data_cost_per_block)

For blake2b it is 128 bytes.

Comment (oxade) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: So for blake2b it'll be` msg.len()/128 * data_cost_per_block` right?

Comment (benr-ml) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: Yes.

@kchalkias wdyt?

Comment (kchalkias) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: yes that's correct, supporting this method to use block size for hash functions Vs single bytes

Comment (kchalkias) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: Yes that's correct, supporting this method to use block size for hash functions Vs single bytes

File: crates/sui-framework/src/natives/crypto/ed25519.rs [hunk 3]
```diff
@@ -1,37 +1,70 @@
...
 use smallvec::smallvec;
-use std::collections::VecDeque;
+use std::{collections::VecDeque, ops::Mul};
 
+#[derive(Clone)]
+pub struct Ed25519VerifyCostParams {
+    /// Base cost for invoking the `ed25519_verify` function
+    pub ed25519_ed25519_verify_cost_base: InternalGas,
+    /// Cost per byte of `msg`
+    pub ed25519_ed25519_verify_msg_cost_per_byte: InternalGas,
```
Comment (kchalkias) [inline crates/sui-framework/src/natives/crypto/ed25519.rs:25]: Same with the discussion we had for hash functions, we are in practice interested in the block size (not per byte) of the internal hash function of ed25519, which is SHA512 (block size for sha512 is 1024bytes)

File: crates/sui-framework/src/natives/crypto/hash.rs (hunk 2 above)
Comment (kchalkias) [reply crates/sui-framework/src/natives/crypto/hash.rs:49]: Re 
> hash_keccak256_data_cost_per_byte - to represent the cost of deserialization between Move and Rust
I'm not sure for that, cause we have to take it into consideration everywhere probably (on native functions) we have to test if hashing operation per block is many many times more expensive than deser cost.

File: crates/sui-framework/src/natives/crypto/ed25519.rs (hunk 3 above)
Comment (benr-ml) [reply crates/sui-framework/src/natives/crypto/ed25519.rs:25]: Block size is 128 here

Comment (oxade) [reply crates/sui-framework/src/natives/crypto/ed25519.rs:25]: addressed now.

File: crates/sui-framework/src/natives/crypto/hash.rs (hunk 2 above)
Comment (oxade) [reply crates/sui-framework/src/natives/crypto/hash.rs:61]: fixed

**Review by benr-ml — COMMENTED** [review]
Looks good! will approve once the cost per block for Ed25519 is in

File: crates/sui-framework/src/natives/crypto/hash.rs [hunk 4]
```diff
@@ -1,50 +1,136 @@
...
     natives::function::NativeResult,
     pop_arg,
     values::{Value, VectorRef},
 };
 use smallvec::smallvec;
-use std::collections::VecDeque;
+use std::{collections::VecDeque, ops::Mul};
+
+const BLAKE_2B256_BLOCK_SIZE: u16 = 128;
+const KECCAK_256_BLOCK_SIZE: u16 = 32;
```
Comment (kchalkias) [inline crates/sui-framework/src/natives/crypto/hash.rs:18]: Are we sure Keccak block size is just 32byte or 136? cc: @benr-ml

Comment (benr-ml) [reply crates/sui-framework/src/natives/crypto/hash.rs:18]: No, my fault, you are right - it should be 136

Comment (oxade) [reply crates/sui-framework/src/natives/crypto/hash.rs]: Got it. Fixed it.

File: crates/sui-framework/src/natives/crypto/hash.rs [hunk 5]
```diff
@@ -1,50 +1,138 @@
...
     natives::function::NativeResult,
     pop_arg,
     values::{Value, VectorRef},
 };
 use smallvec::smallvec;
-use std::collections::VecDeque;
+use std::{collections::VecDeque, ops::Mul};
+
+const BLAKE_2B256_BLOCK_SIZE: u16 = 128;
+const KECCAK_256_BLOCK_SIZE: u16 = 136;
```
Comment (kchalkias) [inline crates/sui-framework/src/natives/crypto/hash.rs:18]: indeed u16 is enough for all of today's popular hash algorithms. Good!

**Review by kchalkias — APPROVED** [review]
LGTM, thanks @oxade

**Comment by oxade** [comment]
> LGTM, thanks @oxade

Thanks for swift turnaround. Will put out PRs for the others shortly.

**Pull request merged**